eo-maven-plugin/src/main/java/org/eolang/maven/fp/Cache.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2016-2024 Objectionary.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.eolang.maven.fp;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Cache directory.
+ * @since 0.41.0
+ */
+public final class Cache {
+    /**
+     * Base cache directory.
+     */
+    private final Path base;
+
+    /**
+     * Cache version.
+     */
+    private final Cache.Version version;
+
+    /**
+     * Relative path of base and version.
+     */
+    private final Path relative;
+
+    /**
+     * Ctor.
+     * @param base Base cache directory
+     * @param version Cache version
+     * @param relative Relative path
+     */
+    public Cache(final Path base, final Cache.Version version, final Path relative) {
+        this.base = base;
+        this.version = version;
+        this.relative = relative;
+    }
+
+    /**
+     * Checks if the current cache directory is cacheable.
+     * The directory is cacheable if cache version is cacheable.
+     * @return True if cacheable
+     */
+    public boolean cacheable() {
+        return this.version.cacheable();
+    }
+
+    /**
+     * Return full absolute path to cache directory.
+     * @return Path to cache directory
+     */
+    public Path path() {
+        return this.base.resolve(this.version.path()).resolve(this.relative);
+    }
+
+    /**
+     * Cache version.
+     * The version consists of two main components:
+     *  1) Base, which is maven-style version string
+     *  2) Hash string, containing git hash.
+     * The two parts define the cacheable location by method {@link #path}.
+     * If base is empty or contains '0.0.0' or 'SNAPSHOT' (see {@link #NOT_CACHEABLE}
+     * versions) the version is considered non-cacheable.
+     * @since 0.41.0
+     */
+    public static class Version {
```

Comment:
@maxonfjvipon why it's an embedded class? We are already in a separate package `fp`, what's the point of placing this class even deeper?